### PR TITLE
[FEAT] Adding `vecdot` implementation

### PR DESCRIPTION
**Pull request by SwayamInSync**
Implementing [NumPy's vecdot](https://numpy.org/devdocs/reference/generated/numpy.vecdot.html) for quaddtype

AI declaration:
Claude is used to extend `test_dot.py` for adding test related `vecdot` and possible edge cases

### CODE REVIEW

**Comment by SwayamInSync** [comment]
This PR also adds some helpers and a new ufunc registering helper function, which will be use for register `matvec` and `vecmat` so I will make those PR after this gets merged

File: src/csrc/umath/matmul.cpp [hunk 1]
```diff
@@ -385,32 +533,26 @@ naive_matmul_strided_loop(PyArrayMethod_Context *context, char *const data[],
...
     if (ufunc == NULL) {
         return -1;
     }
 
     PyArray_DTypeMeta *dtypes[3] = {&QuadPrecDType, &QuadPrecDType, &QuadPrecDType};
 
-#ifndef DISABLE_QUADBLAS
-
     PyType_Slot slots[] = {
             {NPY_METH_resolve_descriptors, (void *)&quad_matmul_resolve_descriptors},
```
Comment (ngoldbaum) [inline src/csrc/umath/matmul.cpp:549]: Claude points out that `quad_matmul_resolve_desctriptors` includes an error message that references matmul. It needs to be generalized to use the ufunc name rather than hardcode it:

https://github.com/numpy/numpy-quaddtype/blob/fc52921783f143a0babf7dd6155fe3acb89d754f/src/csrc/umath/matmul.cpp#L38-L39

Also because `resolve_descriptors` errors out for non-SLEEF backends, longdouble will never actually be able to call the naive loops. This is a pre-existing issue but it's copied here.

File: src/csrc/umath/matmul.cpp [hunk 2]
```diff
@@ -315,6 +315,154 @@ quad_matmul_strided_loop_unaligned(PyArrayMethod_Context *context, char *const d
...
+    npy_intp y_outer_stride = strides[1];
+    npy_intp out_outer_stride = strides[2];
+    npy_intp x_n_stride = strides[3];
+    npy_intp y_n_stride = strides[4];
+
+    QuadPrecDTypeObject *descr = (QuadPrecDTypeObject *)context->descriptors[0];
+    if (descr->backend != BACKEND_SLEEF) {
+        PyErr_SetString(PyExc_NotImplementedError,
+                        "QBLAS-accelerated vecdot only supports SLEEF backend.");
+        return -1;
```
Comment (ngoldbaum) [inline src/csrc/umath/matmul.cpp]: this check is unnecessary because `resolve_descriptors` would have already triggered the same error.

File: src/csrc/umath/matmul.cpp [hunk 3]
```diff
@@ -315,6 +315,154 @@ quad_matmul_strided_loop_unaligned(PyArrayMethod_Context *context, char *const d
...
+    char *y = data[1];
+    char *out = data[2];
+
+    for (npy_intp i = 0; i < N; i++) {
+        Sleef_quad sum = Sleef_cast_from_doubleq1(0.0);
+        for (npy_intp k = 0; k < n; k++) {
+            Sleef_quad a_val, b_val;
+            memcpy(&a_val, x + k * x_n_stride, sizeof(Sleef_quad));
+            memcpy(&b_val, y + k * y_n_stride, sizeof(Sleef_quad));
+            sum = Sleef_fmaq1_u05(a_val, b_val, sum);
```
Comment (ngoldbaum) [inline src/csrc/umath/matmul.cpp]: unaligned matmul does a different thing and calls into `qblas_dot`:

https://github.com/numpy/numpy-quaddtype/blob/fc52921783f143a0babf7dd6155fe3acb89d754f/src/csrc/umath/matmul.cpp#L234-L248

Why the difference? Claude seems to think the `matmul` approach has better numerical accuracy.

File: tests/test_dot.py [hunk 4]
```diff
@@ -689,4 +689,108 @@ def test_dimension_mismatch_matrices(self):
...
+        """vecdot accumulates with quad precision, beating float64 cancellation."""
+        x = create_quad_array([1e20, 1.0, -1e20])
+        y = create_quad_array([0.0, 1.0, 0.0])
+        assert_quad_equal(np.vecdot(x, y), 1.0, atol=1e-25)
+
+    def test_dimension_mismatch(self):
+        x = create_quad_array([1, 2])
+        y = create_quad_array([1, 2, 3])
+        with pytest.raises(ValueError):
+            np.vecdot(x, y)
```
Comment (ngoldbaum) [inline tests/test_dot.py:796]: might be worth adding a test that does vecdot on an empty array, e.g. `np.vecdot(create_quad_array([]), create_quad_array([]))`.

**Review by ngoldbaum — COMMENTED** [review]
Overall looks good, a few comments below.

**Comment by SwayamInSync** [comment]
Thanks @ngoldbaum , I was thinking to re-patch this and #88 after #95 
So if possible we you can next checkout #95 and then I will easily wire the rest ones in.

File: src/csrc/umath/matmul.cpp (hunk 3 above)
Comment (SwayamInSync) [reply src/csrc/umath/matmul.cpp]: Right, This was prior to #95 (and then both qblas_dot and this were same for unaligned access).
Updated it to use the `qblas_dot` as well, since now the operations are also better there + numerical accuracy is similar in both, its more of throughput

File: src/csrc/umath/matmul.cpp (hunk 1 above)
Comment (SwayamInSync) [reply src/csrc/umath/matmul.cpp:549]: Switched to a general error message inside `resolve_descriptors`. I was thinking for `ld`, just the provide the path for naive matmul (or maybe internally through numpy's BLAS dipatch if possible)

File: tests/test_dot.py (hunk 4 above)
Comment (SwayamInSync) [reply tests/test_dot.py:796]: Done!
